yesod-auth/Yesod/Auth.hs

```diff
@@ -110,8 +117,27 @@ class (Yesod master, PathPiece (AuthId master), RenderMessage master FormMessage
     -- destination exists.
     logoutDest :: master -> Route master
 
+    -- | Perform authentication based on the given credentials.
+    --
+    -- Default implementation is in terms of @'getAuthId'@
+    --
+    authenticate :: Creds master -> HandlerT master IO (AuthenticationResult master)
```

Comment:
Not sure what other people think of this, but since `authenticate` gives more information than `getAuthId`, it could be good to recommend that plugin creators implement `authenticate`. (Especially if there's a long-term plan to deprecate `getAuthId`; wasn't sure if that was going to be a thing).